### ISSUE TITLE
Failure to retrieve a row using a numpy.uint64 as index since PyTables 3.2

### ISSUE THREAD

**Issue by 153957** [PyTables/PyTables#475]
Since Pytables 3.2 part of our analysis scripts fail.
It seems we try to use a `numpy.uint64` to retrieve a row from a table, this is apparently no longer supported, it used to work in v3.1. Using a normal `int`, `numpy.uint32` or `numpy.int64` does still work.
Is this intentional? A `uint64` should also be castable to a Python `int`..

With PyTables 3.1:
```python
import tables
data = tables.open_file('muonlab.h5', 'r')
data.root.station_501.events[11]
# (11L, 1433268013, 423254769L, 1433268013423254769L, [316, 417, 6, 5], [2824, 6445, 0, 0], 1.083899974822998, 2.317500114440918, 0.0, 0.0, 995.0, 1180.0, -999.0, -999.0, 1187.5)
data.root.station_501.events[numpy.uint64(11)]
# (11L, 1433268013, 423254769L, 1433268013423254769L, [316, 417, 6, 5], [2824, 6445, 0, 0], 1.083899974822998, 2.317500114440918, 0.0, 0.0, 995.0, 1180.0, -999.0, -999.0, 1187.5)
```

With PyTables 3.2:
```python
import tables
data = tables.open_file('muonlab.h5', 'r')
data.root.station_501.events[11]
# (11L, 1433268013, 423254769L, 1433268013423254769L, [316, 417, 6, 5], [2824, 6445, 0, 0], 1.083899974822998, 2.317500114440918, 0.0, 0.0, 995.0, 1180.0, -999.0, -999.0, 1187.5)
data.root.station_501.events[numpy.uint64(11)]
# ---------------------------------------------------------------------------
# TypeError                                 Traceback (most recent call last)
# <ipython-input-2-64290e3841be> in <module>()
# ----> 1 data.root.station_501.events[numpy.uint64(11)]
# 
# tables/table.pyc in __getitem__(self, key)
#    2112                 key += self.nrows
# -> 2113             (start, stop, step) = self._process_range(key, key + 1, 1)
#    2114             return self.read(start, stop, step)[0]
# 
# tables/leaf.pyc in _process_range(self, start, stop, step, dim, warn_negstep)
#     415 
# --> 416         return slice(start, stop, step).indices(long(nrows))
#     417 
# 
# TypeError: slice indices must be integers or None or have an __index__ method
```

Possibly related #447 

(original issue HiSPARC/sapphire#76)

**Comment by 153957** [comment]
Using an array of uint64 values does work:

```python
data.root.station_501.events[array([4, 5], dtype=numpy.uint64)]
# array([ (4L, 1433270028, 77077861L, 1433270028077077861L, [3891, 2998, -1, -1], [46140, 25131, -1, -1], -999.0, 4.851600170135498, -1.0, -1.0, 495.0, 545.0, -1.0, -1.0, 545.0),
#         (5L, 1433270577, 378396632L, 1433270577378396632L, [562, 490, -1, -1], [3798, 3860, -1, -1], -999.0, 0.745199978351593, -1.0, -1.0, 497.5, 535.0, -1.0, -1.0, 535.0)], dtype=...)
```

**Comment by andreabedini** [comment]
`TypeError: slice indices must be integers or None or have an __index__ method` is saying the right thing. Look at this

```python
In [2]: numpy.uint64(11) + 1
Out[2]: 12.0
```

The slice object becomes `slice(11, 12.0, 1)`, which raises the error (actually the call to `.indices` does).

I am not sure why `numpy.uint64` is behaving in this way. It is still true that

```python
In [5]: numpy.array([11,], dtype=numpy.uint64) + 1
Out[5]: array([12], dtype=uint64)
```

**Comment by 153957** [comment]
So instead of `numpy.uint64(11) + 1` can't it do `int(numpy.uint64(11)) + 1` (if type == numpy.uint64)?

**Comment by andreabedini** [comment]
@153957 yes, there're many ways we can work around this. I was trying to understand why it fails before fixing it.
1) `numpy.integer` derives from `numpy.integer`, we should test against that to be general.
2) using `int()` we would introduce an unnecessary conversion.
3) there should be a reason behind `numpy.uint64(11) + 1 = 12.0` although I can't see it now

further insight or patches are welcome!

**Comment by 153957** [comment]
Perhaps related: https://github.com/PyTables/PyTables/commit/4f8152daca445826aea5199169ae7bd37b37ca86
No variable into which it can be upcasted.

**Comment by andreabedini** [comment]
well spotted @153957! It's still not clear to me what the deal is.

**Comment by FrancescAlted** [comment]
I would say that problem should be in some new behavior introduced in recent versions of NumPy.  At any rate, this PR looks promising.

**Comment by andreabedini** [comment]
@FrancescAlted I found it!

PyTables behaviour changed in 4f9c5d1. After a long discussion in #195, we ignored your [wise words](https://github.com/PyTables/PyTables/commit/4f9c5d1581a42e2cf91ae1edd88d98d7894f255c#diff-aafa3b17e8a6f3ecec60ebc15269cd8bL404) and we removed the custom substitute for `slice().indices`. The removed code would cast all its arguments to longs. With the new version that cast disappeared and @153957's code broke :)

Now, as I mentioned in #496 I am afraid casting to int might hurt users on platforms where a Python int is 32 bit. 3.2.2 is meant to be a bug fix release so I don't want to screw up too much.
